cps.nim

```
@@ -895,70 +858,38 @@ proc cpsXfrmProc(T: NimNode, n: NimNode): NimNode =
   # spamming the developers
   debug(".cps.", result, Transformed)
 
-proc cpsXfrm(T: NimNode, n: NimNode): NimNode =
-  # Perform CPS transformation on a NimNode. This can be a single
-  # proc, or a top level stmtList.
-  case n.kind
-  of nnkProcDef:
-    result = cpsXfrmProc(T, n)
-  of nnkStmtList:
-    result = copyNimNode n
-    for nc in n.items:
-      result.add cpsXfrm(T, nc)
-  else:
-    result = copy n
-  result = workaroundRewrites(result)
-
 macro cps*(T: typed, n: typed): untyped =
-  # I hate doing stuff inside macros, call the proc to do the work
+  ## This is the .cps. macro performing the proc transformation
   when defined(nimdoc):
     result = n
   else:
-    result = cpsXfrm(T, n)
+    result = cpsXfrmProc(T, n)
 
 macro cpsMagic*(n: untyped): untyped =
   ## upgrade cps primitives to generate errors out of context
   ## and take continuations as input inside {.cps.} blocks
   expectKind(n, nnkProcDef)
-  result = newStmtList()
 
-  # ensure that .cpsCall. is added to the copies of the proc
+  # Add .cpsCall. pragma to the proc
   n.addPragma ident"cpsCall"
 
-  # create a version of the proc that pukes outside of cps context
+  # create a Nim-land version of the proc that throws an exception when called
+  # from outside of CPS-land. XXX We also need this to keep the nim compiler
+  # happy, otherwise we get a type mismatch on call. Do we need a special name
+  # for this type of proc?
```

Input on the above appreciated

---

Honestly, I don't think it should throw an exception outside; the library author may have a legitimate reason to use a procedure like `sleep` or `enqueue` on a continuation.  What about creating a `warning` outside of `release` builds?

---

The version outside of CPS do nothing, and doesn't even take a continuation.

---

Right, it only exists so Nim is able to compile the code in the first place